common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -77,57 +92,142 @@ func (s *localStoreCertProvider) FetchServerCertificate() (*tls.Certificate, err
+	}
+
+	keyBytes, err := base64.StdEncoding.DecodeString(s.tlsSettings.Server.KeyData)
+	if err != nil {
+		return nil, fmt.Errorf("TLS certificate private key could not be decoded: %w", err)
+	}
+
+	serverCert, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, fmt.Errorf("loading server tls certificate failed: %v", err)
+	}
+
+	s.serverCert = &serverCert
+	return s.serverCert, nil
+}
+
 func (s *localStoreCertProvider) FetchClientCAs() (*x509.CertPool, error) {
-	if s.tlsSettings.Server.ClientCAFiles == nil {
+	if !s.tlsSettings.Server.InlineData && len(s.tlsSettings.Server.ClientCAFiles) == 0 {
+		return nil, nil
+	}
+	if s.tlsSettings.Server.InlineData && len(s.tlsSettings.Server.ClientCaData) == 0 {
 		return nil, nil
 	}
 
-	// Check under a read lock first
 	s.RLock()
 	if s.clientCAs != nil {
 		defer s.RUnlock()
 		return s.clientCAs, nil
 	}
-	// Not found, manually unlock read lock and move to write lock
+
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
+	if s.clientCAs != nil {
+		return s.clientCAs, nil
+	}
+
+	if s.tlsSettings.Server.InlineData {
+		return s.fetchClientCAsFromInline()
+	}
+
+	return s.fetchClientCAsFromFiles()
+}
+
+func (s *localStoreCertProvider) fetchClientCAsFromFiles() (*x509.CertPool, error) {
+	if len(s.tlsSettings.Server.ClientCAFiles) == 0 {
+		return nil, nil
+	}
+
 	var clientCaPool *x509.CertPool
-	if len(s.tlsSettings.Server.ClientCAFiles) > 0 {
-		var err error
-		clientCaPool, err = buildCAPool(s.tlsSettings.Server.ClientCAFiles)
-		if err != nil {
-			return nil, err
-		}
+	var err error
+	clientCaPool, err = buildCAPool(s.tlsSettings.Server.ClientCAFiles)
+	if err != nil {
+		return nil, err
+	}
+
+	s.clientCAs = clientCaPool
+	return s.clientCAs, nil
+}
+
+func (s *localStoreCertProvider) fetchClientCAsFromInline() (*x509.CertPool, error) {
+	if len(s.tlsSettings.Server.ClientCaData) == 0 {
+		return nil, nil
+	}
+
+	var clientCaPool *x509.CertPool
+	var err error
+	clientCaPool, err = buildCAPoolFromInline(s.tlsSettings.Server.ClientCaData)
+	if err != nil {
+		return nil, err
 	}
 
 	s.clientCAs = clientCaPool
 	return s.clientCAs, nil
 }
 
 func (s *localStoreCertProvider) FetchServerRootCAsForClient() (*x509.CertPool, error) {
-	if s.tlsSettings.Client.RootCAFiles == nil {
+	if !s.tlsSettings.Client.InlineData && len(s.tlsSettings.Client.RootCAFiles) == 0 {
+		return nil, nil
+	}
+	if s.tlsSettings.Client.InlineData && len(s.tlsSettings.Client.RootCAData) == 0 {
 		return nil, nil
 	}
-	// Check under a read lock first
+
 	s.RLock()
 	if s.serverCAs != nil {
 		defer s.RUnlock()
 		return s.clientCAs, nil
 	}
-	// Not found, manually unlock read lock and move to write lock
+
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
+	if s.tlsSettings.Client.InlineData {
+		return s.fetchServerRootCAsForClientFromInline()
+	}
+
+	return s.fetchServerRootCAsForClientFromFiles()
+}
+
+func (s *localStoreCertProvider) fetchServerRootCAsForClientFromFiles() (*x509.CertPool, error) {
```

Comment:
Same suggestion here - avoid code duplication between `fetchServerRootCAsForClientFromFiles()` and `fetchServerRootCAsForClientFromInline()` and don't give them access to `s`.

Reply:
removed helper function